sklearn_helpers/search_param.py

```diff
@@ -0,0 +1,119 @@
+
+
+
+
+from sklearn.model_selection import ParameterGrid as grid
+import random
+import numpy as np
+
+
+def fl_search(fun,params:dict,n_iter:int)->dict:
+    """
+    This function is to actively search best parameters for one model, should be better than grid search.
+    parameter space should be discrete and monotonic.
+    :param params: The parameters space to search, with keys as parameters and values are list of candidate values
+    :return: A tuple with first one is the dictionary for the best parameters, and second one is a list as the best score tracking
+    """
+
+
+    keys=list(params.keys())
+
+    num_points={}
+    for key in keys:
+        num_points[key]=len(params[key])
+        params[key].sort()
+
+    lower_point, upper_point=_init_upper_lower_points(keys=keys,num_points=num_points)
+    move_up={}
+    tracking=[]
+
+    i=0
+    while i<n_iter:
+        # find the move direction for next round
+        score,move_up= _find_move_direction(fun=fun,keys=keys,params=params,upper_point=upper_point,
+                                          lower_point=lower_point,move_up=move_up)
+
+        # Track the score for the optimization
+        tracking.append(score)
+        param={}
+        for key in keys:
+            if move_up[key]:
+                param[key] = params[key][upper_point[key]]
+            else:
+                param[key] = params[key][lower_point[key]]
+
+        # Reset the lower_point and upper_point based move direction
+        lower_point, upper_point = _reset_upper_lower_points(keys=keys, move_up=move_up,num_points=num_points,
+                                                             upper_point=upper_point,lower_point=lower_point)
+        i+=1
+
+
+    return (param,tracking)
+
+def _find_move_direction(fun,keys:list,params:dict,upper_point:dict,lower_point:dict,move_up:dict)->tuple:
+    """
+    This function is to calculate the best combination of upper_point and lower_point. The best one decide the moving
+    direction for each parameter. The best score should be also stored.
+    :param fun: One callable function to optimize, with inputs as the keys.
+    :param keys: A list of strings, from the keys of params
+    :param params: A dictionary containing all parameter space to search
+    :param upper_point: A dictionary with keyse from params, values containing upper point index for each parameter
+    :param lower_point: A dictionary with keyse from params, values containing lower point index for each parameter
+    :param move_up: A dictionary with keyse from params, values are logic ones describing move up or down.
+    :return: A tuple, with the first one is best score, and the second one is the moving direction move_up
+    """
+    best_score=np.Inf
+    move_space={}
+    for key in keys:
+        move_space[key]=[False,True]
+
+    for move in grid(move_space):
```

Comment:
We can clean this up with 
```
moves = grid({key: (lower_point[key], upper_point[key]) for key in keys})
scores = {param: fun(param) for param in moves}
return max(scores.items(), key=operator.itemgetter(1))
```
We of course need to import at operator to do this, and the function now returns the score and the parameter dictionary that produced the score, but this seems like it's ok, because it looks like the code just uses the `move_up` dictionary to produce the parameter dictionary.